Makefile doesn't handle filepaths with spaces, and builds fail on filepaths it does accept

Error message using WSL trying to make in a filepath with spaces:

```
agratton@DESKTOP-O12R4VM:/mnt/c/Users/gratt/Documents/University/Design Teams/UW Mars Rover/MarsRover2020-firmware$ make APP=arm TARGET=arm
make[1]: Entering directory '/mnt/c/Users/gratt/Documents/University/Design Teams/UW Mars Rover/MarsRover2020-firmware/build'
make[1]: /mnt/c/Users/gratt/Documents/University/Design: No such file or directory
make[1]: *** No rule to make target '/mnt/c/Users/gratt/Documents/University/Design'.  Stop.
make[1]: Leaving directory '/mnt/c/Users/gratt/Documents/University/Design Teams/UW Mars Rover/MarsRover2020-firmware/build'
makefile:56: recipe for target 'all' failed
make: *** [all] Error 2

```

When using a filepath that has no spaces, build fails to compile with many repetitions of this type of error (tested using test-blinky and arm apps for target arm):
```
/usr/include/newlib/c++/6.3.1/bits/stl_algo.h:5536:3: error: template with C linkage
   template<typename _ForwardIterator>
   ^~~~~~~~
```

--- a/makefile
+++ b/makefile
@@ -27,6 +27,16 @@ LAST_BOARD_TARGET := $(shell cat $(BUILD_FOLDER)/LAST_BOARD_TARGET 2>/dev/null)#
 # Utility
 ###############################################################################
 
+# Whitespace definitions
+EMPTY :=
+SPACE := ${EMPTY} ${EMPTY}
+${SPACE} := ${SPACE}
+
+define \n
+
+
+endef
+
 # Cross-platform commands
 ifeq ($(shell echo $$OS),$$OS)
     MAKE_DIR = if not exist "$(1)" mkdir "$(1)"
@@ -37,19 +47,14 @@ else
     RM_DIR = '$(SHELL)' -c "rm -rf \"$(1)\""
     RM_FILE_TYPE = '$(SHELL)' -c "find $(1) -name \"$(2)\" -delete -print"
 endif
-
-# Whitespace definitions
-null :=
-space := ${null} ${null}
-${space} := ${space}
-
-define \n
-
-
-endef
+                        
+# Other Convenience Commands
+REPLACE_SPACES = $(subst $(SPACE),?,$(1))
+RESTORE_SPACES = $(subst ?,$(SPACE),$(1))
+NOTDIR_PATH_WITH_SPACES = $(call RESTORE_SPACES,$(notdir $(call REPLACE_SPACES,$(1))))
 
 # If not in build directory, move to the build directory
-ifneq ($(BUILD_FOLDER),$(notdir $(CURDIR)))
+ifneq ($(BUILD_FOLDER),$(call NOTDIR_PATH_WITH_SPACES,$(CURDIR)))
 
 .PHONY: $(BUILD_FOLDER) clean
 all:
@@ -62,10 +67,10 @@ ifeq (,$(findstring $(TARGET), $(patsubst $(TARGETS_FOLDER)/%/,%,$(sort $(dir $(
 endif
 ifneq ($(TARGET),$(LAST_BOARD_TARGET))
 	$(info New TARGET detected. Recompiling all files.)
-	$(call RM_DIR,$(BUILD_FOLDER)/$(OBJS_FOLDER))
+	+@$(call RM_DIR,$(BUILD_FOLDER)/$(OBJS_FOLDER))
 	+@echo $(TARGET) > $(BUILD_FOLDER)/LAST_BOARD_TARGET
 endif
-	+@$(MAKE) --directory=$(BUILD_FOLDER) --file=$(abspath $(lastword $(MAKEFILE_LIST))) $(MAKECMDGOALS)
+	+@$(MAKE) --directory='$(BUILD_FOLDER)' --file='$(abspath $(lastword $(MAKEFILE_LIST)))' $(MAKECMDGOALS)
 
 $(BUILD_FOLDER): all
 

